File: backend/src/main/java/org/example/service/OrderService.java
package org.example.service;

import org.example.dto.request.OrderRequest;
import org.example.dto.response.ApiResponse;
import org.example.dto.response.OrderResponse;
import org.example.entity.*;
import org.example.exception.BadRequestException;
import org.example.exception.ResourceNotFoundException;
import org.example.repository.*;
import org.example.mapper.OrderMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import jakarta.validation.Valid;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class OrderService {

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private OrderDetailRepository orderDetailRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private CartRepository cartRepository;

    @Autowired
    private CartItemRepository cartItemRepository;

    @Autowired
    private OrderMapper orderMapper;

    /**
     * Lấy danh sách đơn hàng của user
     */
    public List<OrderResponse> getMyOrders(String userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        List<Order> orders = orderRepository.findByUser(user);
        
        return orders.stream()
            .map(this::toOrderResponseWithDetails)
            .collect(Collectors.toList());
    }

    /**
     * Lấy chi tiết đơn hàng
     */
    public OrderResponse getOrderById(String userId, String orderId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        Order order = orderRepository.findById(orderId)
            .orElseThrow(() -> new ResourceNotFoundException("Order not found"));

        // Verify ownership
        if (!order.getUser().getId().equals(userId)) {
            throw new BadRequestException("Bạn không có quyền xem đơn hàng này");
        }

        return toOrderResponseWithDetails(order);
    }

    /**
     * Tạo đơn hàng từ giỏ hàng
     */
    public ApiResponse<OrderResponse> createOrder(String userId, @Valid OrderRequest orderRequest) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        Cart cart = cartRepository.findByUser(user)
            .orElseThrow(() -> new ResourceNotFoundException("Cart not found"));

        List<CartItem> cartItems = cartItemRepository.findByCart(cart);
        
        if (cartItems.isEmpty()) {
            throw new BadRequestException("Giỏ hàng trống");
        }

        // Tính tổng tiền
        BigDecimal totalAmount = cartItems.stream()
            .map(item -> item.getCar().getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
            .reduce(BigDecimal.ZERO, BigDecimal::add);

        // Tạo đơn hàng
        Order order = new Order();
        order.setUser(user);
        order.setTotalAmount(totalAmount);
        order.setStatus(Order.OrderStatus.PENDING);
        order.setDeliveryAddress(orderRequest.getDeliveryAddress());
        order.setPaymentMethod(orderRequest.getPaymentMethod());
        order.setNotes(orderRequest.getNotes());
        order.setOrderDate(LocalDateTime.now());

        Order savedOrder = orderRepository.save(order);

        // Tạo chi tiết đơn hàng
        for (CartItem cartItem : cartItems) {
            OrderDetail orderDetail = new OrderDetail();
            orderDetail.setOrder(savedOrder);
            orderDetail.setCar(cartItem.getCar());
            orderDetail.setQuantity(cartItem.getQuantity());
            orderDetail.setUnitPrice(cartItem.getCar().getPrice());
            orderDetailRepository.save(orderDetail);
        }

        // Xóa giỏ hàng
        cartItemRepository.deleteAll(cartItems);

        return ApiResponse.<OrderResponse>builder()
            .success(true)
            .message("Đặt hàng thành công")
            .data(toOrderResponseWithDetails(savedOrder))
            .timestamp(LocalDateTime.now())
            .build();
    }

    /**
     * Hủy đơn hàng
     */
    public ApiResponse<String> cancelOrder(String userId, String orderId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        Order order = orderRepository.findById(orderId)
            .orElseThrow(() -> new ResourceNotFoundException("Order not found"));

        // Verify ownership
        if (!order.getUser().getId().equals(userId)) {
            throw new BadRequestException("Bạn không có quyền hủy đơn hàng này");
        }

        // Chỉ cho phép hủy đơn hàng PENDING
        if (order.getStatus() != Order.OrderStatus.PENDING) {
            throw new BadRequestException("Chỉ có thể hủy đơn hàng đang chờ xác nhận");
        }

        order.setStatus(Order.OrderStatus.CANCELLED);
        orderRepository.save(order);

        return ApiResponse.<String>builder()
            .success(true)
            .message("Hủy đơn hàng thành công")
            .data("Order cancelled")
            .timestamp(LocalDateTime.now())
            .build();
    }

    /**
     * Lấy đơn hàng theo trạng thái
     */
    public List<OrderResponse> getOrdersByStatus(String userId, Order.OrderStatus status) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        List<Order> orders = orderRepository.findByUserAndStatus(user, status);
        
        return orders.stream()
            .map(this::toOrderResponseWithDetails)
            .collect(Collectors.toList());
    }

    /**
     * [Admin] Lấy tất cả đơn hàng
     */
    public Page<OrderResponse> getAllOrders(int page, int size, String sortBy, String sortDir) {
        Sort sort = sortDir.equalsIgnoreCase("desc") ? 
            Sort.by(sortBy).descending() : Sort.by(sortBy).ascending();
        
        Pageable pageable = PageRequest.of(page, size, sort);
        Page<Order> orders = orderRepository.findAll(pageable);
        
        return orders.map(this::toOrderResponseWithDetails);
    }

    /**
     * [Admin] Cập nhật trạng thái đơn hàng
     */
    public ApiResponse<OrderResponse> updateOrderStatus(String orderId, Order.OrderStatus newStatus) {
        Order order = orderRepository.findById(orderId)
            .orElseThrow(() -> new ResourceNotFoundException("Order not found"));

        order.setStatus(newStatus);
        Order updatedOrder = orderRepository.save(order);

        return ApiResponse.<OrderResponse>builder()
            .success(true)
            .message("Cập nhật trạng thái thành công")
            .data(toOrderResponseWithDetails(updatedOrder))
            .timestamp(LocalDateTime.now())
            .build();
    }

    private OrderResponse toOrderResponseWithDetails(Order order) {
        OrderResponse response = orderMapper.toOrderResponse(order);
        
        List<OrderDetail> orderDetails = orderDetailRepository.findByOrder(order);
        List<OrderResponse.OrderDetailResponse> detailResponses = orderDetails.stream()
            .map(orderMapper::toOrderDetailResponse)
            .collect(Collectors.toList());
        
        response.setOrderDetails(detailResponses);
        
        return response;
    }
}

File: frontend/src/services/orderService.js
import { ApiService } from './api.js';

/**
 * Order service - Quản lý đơn hàng
 * @typedef {import('../types/order.types.js').Order} Order
 * @typedef {import('../types/order.types.js').CreateOrderRequest} CreateOrderRequest
 * @typedef {import('../types/order.types.js').OrderFilterParams} OrderFilterParams
 */
class OrderService extends ApiService {
  constructor() {
    super('/orders');
  }

  /**
   * Lấy danh sách đơn hàng của user
   * @returns {Promise<Object>} Danh sách đơn hàng
   */
  async getMyOrders() {
    return await this.get('');
  }

  /**
   * Lấy chi tiết đơn hàng
   * @param {string} orderId - Order ID
   * @returns {Promise<Object>} Chi tiết đơn hàng
   */
  async getOrderById(orderId) {
    return await this.get(`/${orderId}`);
  }

  /**
   * Tạo đơn hàng mới
   * @param {CreateOrderRequest} orderData - Dữ liệu đơn hàng
   * @returns {Promise<Object>} Đơn hàng đã tạo
   */
  async createOrder(orderData) {
    return await this.post('', orderData);
  }

  /**
   * Hủy đơn hàng
   * @param {string} orderId - Order ID
   * @returns {Promise<Object>} Kết quả hủy
   */
  async cancelOrder(orderId) {
    return await this.delete(`/${orderId}`);
  }

  /**
   * Lấy đơn hàng theo trạng thái
   * @param {string} status - Trạng thái
   * @returns {Promise<Object>} Danh sách đơn hàng
   */
  async getOrdersByStatus(status) {
    return await this.get(`/status/${status}`);
  }

  // ==================== ADMIN METHODS ====================

  /**
   * [Admin] Lấy tất cả đơn hàng
   * @param {OrderFilterParams} params - Tham số lọc
   * @returns {Promise<Object>} Danh sách đơn hàng
   */
  async getAllOrders(params = {}) {
    const queryParams = {
      page: params.page || 0,
      size: params.size || 10,
      sortBy: params.sortBy || 'createdAt',
      sortDir: params.sortDir || 'desc'
    };

    if (params.status) queryParams.status = params.status;
    if (params.fromDate) queryParams.fromDate = params.fromDate;
    if (params.toDate) queryParams.toDate = params.toDate;

    return await this.get('/all', { params: queryParams });
  }

  /**
   * [Admin] Cập nhật trạng thái đơn hàng
   * @param {string} orderId - Order ID
   * @param {string} status - Trạng thái mới
   * @returns {Promise<Object>} Đơn hàng đã cập nhật
   */
  async updateOrderStatus(orderId, status) {
    return await this.patch(`/${orderId}/status`, null, {
      params: { status }
    });
  }

  /**
   * Tính tổng tiền đơn hàng
   * @param {Array} items - Danh sách sản phẩm
   * @returns {number} Tổng tiền
   */
  calculateTotal(items) {
    if (!items || !Array.isArray(items)) return 0;
    return items.reduce((sum, item) => {
      return sum + (item.price || 0) * (item.quantity || 1);
    }, 0);
  }

  /**
   * Lấy thống kê đơn hàng của user
   * @returns {Promise<Object>} Thống kê
   */
  async getOrderStats() {
    try {
      const response = await this.getMyOrders();
      const orders = response.data || [];

      const stats = {
        total: orders.length,
        pending: 0,
        confirmed: 0,
        processing: 0,
        shipped: 0,
        delivered: 0,
        cancelled: 0,
        totalSpent: 0
      };

      orders.forEach(order => {
        const status = order.status?.toLowerCase();
        if (stats[status] !== undefined) {
          stats[status]++;
        }
        if (order.status !== 'CANCELLED') {
          stats.totalSpent += order.totalAmount || 0;
        }
      });

      return {
        success: true,
        data: stats
      };
    } catch (error) {
      throw error;
    }
  }
}

export default new OrderService();
